### DISCUSSION TITLE
Differing behavior for Java registry url configuration

### DISCUSSION THREAD

**Discussion by setchy** [renovatebot/renovate#16951] [Request Help] [answered]
I've noticed the following odd behavior and am curious as to why Java `registryUrls` can only be defined using a `packageRules` and not via the [java configuration object](https://docs.renovatebot.com/configuration-options/#java)

For example, the following configuration appears to have zero influence on the registries used for Java (maven/gradle)
```
"java": {
    "registryUrls": [
      "https://foo"
    ]
  },
```

Instead, either of the following `packageRules` examples have appeared to work
```
 "packageRules": [
    {
      "matchDatasources": [
        "maven",
        "gradle"
      ],
      "registryUrls": [
        "https://foo"
      ]
    }
]
```

or

```
"java": {
    "packageRules": [
      {
        "matchDatasources": [
          "maven",
          "gradle"
        ],
        "registryUrls": [
          "https://foo"
        ]
      },
  ]
}
```

Other managers on the other hand, such as `dotnet` or `docker`, do seem to support setting `registryUrls` via their configuration object

**Comment by viceice** [comment, marked as answer]
It's because datasources don't support the `languages` config options and not all managers are reporting the `language`, so the `java` object isn't applied.

`languages` are deprecated and will be changed to some kind of tags soon. This needs to be done, as some managers support different languages. 

We suggest to always use `packagesRules`, as they have a defined order.

**Reply by setchy** [reply]
Thanks @viceice - good to know that `languages` is deprecated, I hadn't read this before.

So you are recommending to use `pacakgeRules` with either `matchDatasources` or `matchManagers`?

In my sandbox, this indeed does work for `maven` and `gradle`, but does not work for `nuget` and a few others that I cannot recall offhand

**Reply by viceice** [reply]
yes `packagesRules` should work for all cases. PLease open new discussion for any rules that don't work as expected.
